NitroxClient/Communication/ClientBridge.cs

```diff
@@ -0,0 +1,249 @@
+
+namespace NitroxClient.Communication
+{
+    //This is meant to be a drop-in replacement for the old PacketSender. It was re-written in accordance with the Bridge design pattern to clarify intent.
+    //This pattern should give us greater flexibility in how we bridge the connection between the server and the client moving forward. 
+    //Example: We can refactor this implementation into a ReservableClientBridge, allowing us to implement different types of client bridges depending on specific requirements.
+    public class ClientBridge : IClientBridge
+    {
+        private readonly HashSet<Type> suppressedPacketsTypes;
+        private IClient serverClient;
+        private string correlationId;
+
+        public ClientBridgeState CurrentState { get; private set; }
+        public string ReservationKey { get; private set; }
+        public ReservationRejectionReason ReservationRejectionReason { get; private set; }
+
+        //Eww...
+        public string PlayerId { get; private set; }
+
+        public ClientBridge(IClient serverClient)
+        {
+            Log.Info("Initializing ClientBridge...");
+            suppressedPacketsTypes = new HashSet<Type>();
+            this.serverClient = serverClient;
+            correlationId = GenerateCorrelationId();
+
+            CurrentState = ClientBridgeState.Disconnected;
+            ReservationKey = null;
+            ReservationRejectionReason = ReservationRejectionReason.None;
+        }
+
+        public void Connect(string ipAddress, string playerName)
+        {
+            try
+            {
+                if (serverClient.IsConnected == false)
+                {
+                    serverClient.Start(ipAddress);
+
+                    if (serverClient.IsConnected == false)
+                    {
+                        throw new ClientConnectionFailedException("The client failed to indicate a Connected status without providing a reason for why it cannot reach the server.");
+                    }
+                }
+            }
+            catch (ClientConnectionFailedException ex)
+            {
+                TryStopClient();
+                CurrentState = ClientBridgeState.Failed;
+                throw ex;
+            }
+            catch (Exception ex)
+            {
+                TryStopClient();
+                CurrentState = ClientBridgeState.Failed;
+                throw new ClientConnectionFailedException("The connect operation failed for an unknown reason. Investigate the inner exception for additional details.", ex);
+            }
+
+            RequestReservation(playerName);
+        }
+
+        public void Disconnect()
+        {
+            TryStopClient();
+            CurrentState = ClientBridgeState.Disconnected;
+        }
+
+        public void ConfirmReservation(string correlationId, string reservationKey)
+        {
+            try
+            {
+                if (CurrentState != ClientBridgeState.WaitingForRerservation)
```

Comment:
Is it possible to refactor this enum based state machine into something more object oriented?  The current approach makes a larger file and peppers state logic everywhere.  I was doing some brainstorming and thinking that states can chain themselves together in memory like a linked list:

```
public class AwaitingReservation : IClientBridgeState {

   private IClientBridgeState connectedState;
   private IClientBridgeState shutdownClientState;

   public IClientBridgeState Next(StateContext stateContext) {
	    assert(context is ReservationResultContext); 

        if(context.RejectionReason == None) {
	        return SuccessfulReservation(context);
        } else {
	        return shutdownClientState;
        }
   }

   private IClientBridgeState SuccessfulReservation(ReservationResultContext context) {
        ValidateCorrelationId(correlationId);
        ValidateReservationKey(reservationKey);
        ConfirmCorrelationId(correlationId);

	     var packet = new ClaimPlayerSlotReservation(correlationId, ReservationKey);
        serverClient.Send(packet);

        return connectedState;
   }

}
```

Not a blocker at all - feel free to take it,  leave it, or just experiment.

Reply:
That's certainly in the master plan, but I figured that would come whenever I refactored that logic into a ReservationManager that sits outside of this construct. This was honestly the first pass through, and did not want to spend any additional iterations on it until we had at least talked about addressing the way dependencies are being managed.